### PR TITLE
Feature/unit testing

### PR DESCRIPTION
**Pull request by Brennan1994**
Adds unit tests for paired data and some refactoring to the paired data class

### CODE REVIEW

File: fda-model-test/PairedDataTest.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        [SetUp]
+        public void Setup()
+        {
+ 
+        }
+
+        [Test]
+        public void TestGetXvals()
+        { 
+            Assert.AreEqual(xs,myPairedData.Xvals);
```
Comment (HenryGeorgist) [inline fda-model-test/PairedDataTest.cs]: This is a great test to get something started - but testing access modifiers is probably not super necessary. We will be blue in the face writing this level of testing all the time. We need to test the mission critical code like integrate, multiply and compose. (I am also advocating we seek to protect the xs and ys from public access)

File: fda-model/paireddata/PairedData.cs [hunk 2]
```diff
@@ -21,25 +21,40 @@ public class PairedData: IPairedData
...
+                //This is the next LARGER value.
+                idx = ~idx;
+
+                if(idx == Xvals.Count)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+                if(idx == 0)
+                {
+                    throw new IndexOutOfRangeException();
```
Comment (HenryGeorgist) [inline fda-model/paireddata/PairedData.cs:40]: please no exception here.

File: fda-model-test/PairedDataTest.cs [hunk 3]
```diff
@@ -1,29 +1,44 @@
...
-        { 
-            Assert.AreEqual(xs,myPairedData.Xvals);
+        public void GetYvals()
+        {
+            Assert.AreEqual(Stages, myRatingCurve.Yvals);
+        }
+        [Test]
+        public void FInterpolatesCorrect()
+        {
+            double expected = 565.5; //linear interp of ys when x=7500
```
Comment (HenryGeorgist) [inline fda-model-test/PairedDataTest.cs:30]: i would use a more clear case if your x values and y values are 1 to 1 we should get back equivalent values, if they are 1 to 2 we get back double our intput. those are really easy to test.

Comment (HenryGeorgist) [reply fda-model-test/PairedDataTest.cs:30]: But this is great this is an excellent test we need to verify. We need to test more cases, over, under, exact values, inbetween values.

File: fda-model-test/PairedDataTest.cs [hunk 4]
```diff
@@ -1,29 +1,44 @@
...
         }
         [Test]
-        public void TestGetYvals()
+        public void ComposeComposes()
         {
-            Assert.AreEqual(ys, myPairedData.Yvals);
+            IPairedData expected = new PairedData(Probabilities, Stages); //Probability-Stage
+            IPairedData actual = myRatingCurve.compose(myFlowFrequencyCurve);
+            Assert.AreEqual(expected.Xvals,actual.Xvals);
+            Assert.AreEqual(expected.Yvals, actual.Yvals);
```
Comment (HenryGeorgist) [inline fda-model-test/PairedDataTest.cs:40]: great test.

**Review by HenryGeorgist — COMMENTED** [review]
I would like to talk with you about these changes, generally they are phenominal, but i am really uncomfortable with the way you changed the paired data class, i was about to refactor it to be more protected.

File: fda-model-test/PairedDataTest.cs (hunk 1 above)
Comment (Brennan1994) [reply fda-model-test/PairedDataTest.cs]: I'm good with that. I'd be interested in deleting this right back out with that in mind. I don't want to add noise for no good reason.

File: fda-model/paireddata/PairedData.cs [hunk 5]
```diff
@@ -21,25 +21,40 @@ public class PairedData: IPairedData
...
                 return Yvals[idx];
             }
+            else
+            {
+                //This is the next LARGER value.
+                idx = ~idx;
+
+                if(idx == Xvals.Count)
+                {
+                    throw new IndexOutOfRangeException();
```
[the remark this answers is not in the document]
Comment (Brennan1994) [reply fda-model/paireddata/PairedData.cs:36]: I can make that change for sure, Do we need to be able to do some sort of validation in the model though?

**Pull request closed without merging**